src/client/producer/broadcast.rs

```diff
@@ -1,76 +1,101 @@
-use futures::future::{Future, FutureExt};
-use pin_project_lite::pin_project;
-use std::pin::Pin;
-use std::task::{Context, Poll};
-use tokio::sync::oneshot;
+use parking_lot::Mutex;
+use std::sync::Arc;
+use thiserror::Error;
+use tokio::sync::Notify;
 use tracing::warn;
 
-/// A broadcast channel that can send a single value
-///
-/// - Receivers can be created with `BroadcastOnce::receiver`
-/// - The value can be produced with `BroadcastOnce::broadcast`
-pub struct BroadcastOnce<T: Clone> {
-    receiver: futures::future::Shared<ReceiveFut<T>>,
-    sender: oneshot::Sender<T>,
+#[derive(Debug, Error, Clone, PartialEq)]
+pub enum Error {
+    #[error("BroadcastOnce dropped")]
+    Dropped,
 }
 
-impl<T: Clone> std::fmt::Debug for BroadcastOnce<T> {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "ResultSlot")
-    }
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+/// A broadcast channel that can send a single value
+///
+/// - Receivers can be created with [`BroadcastOnce::receiver`]
+/// - The value can be produced with [`BroadcastOnce::broadcast`]
+#[derive(Debug)]
+pub struct BroadcastOnce<T> {
```

Comment:
As described in #111 `Shared::peak` has some quirks when used with interior mutable futures, these can be worked around, but in the interest of keeping things simple I just switch to using Notify and a shared Option